File: go-backend/internal/util/jwt.go
package util

import (
	"fmt"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

var jwtSecret string
var jwtExpiry string
var jwtRefreshExpiry string

func InitJWT(secret, expiry, refreshExpiry string) {
	jwtSecret = secret
	jwtExpiry = expiry
	jwtRefreshExpiry = refreshExpiry
}

type Claims struct {
	Email string `json:"email"`
	Type  string `json:"type"`
	jwt.RegisteredClaims
}

func GenerateAccessToken(email string) (string, error) {
	duration, err := time.ParseDuration(jwtExpiry)
	if err != nil {
		return "", fmt.Errorf("invalid JWT_EXPIRY duration: %w", err)
	}

	claims := Claims{
		Email: email,
		Type:  "access",
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(duration)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(jwtSecret))
}

func GenerateRefreshToken(email string) (string, error) {
	duration, err := time.ParseDuration(jwtRefreshExpiry)
	if err != nil {
		return "", fmt.Errorf("invalid JWT_REFRESH_EXPIRY duration: %w", err)
	}

	claims := Claims{
		Email: email,
		Type:  "refresh",
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(duration)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(jwtSecret))
}

func VerifyToken(tokenString string) (*Claims, error) {
	claims := &Claims{}
	token, err := jwt.ParseWithClaims(tokenString, claims, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return []byte(jwtSecret), nil
	})

	if err != nil {
		return nil, err
	}

	if !token.Valid {
		return nil, fmt.Errorf("invalid token")
	}

	return claims, nil
}

func DecodeToken(tokenString string) *Claims {
	claims, err := VerifyToken(tokenString)
	if err != nil {
		return nil
	}
	return claims
}

File: backend/src/utils/jwt.ts
import jwt from 'jsonwebtoken';

const JWT_SECRET = process.env.JWT_SECRET || 'pos-super-secret-jwt-key-change-in-production';
const JWT_EXPIRY = process.env.JWT_EXPIRY || '30d';
const JWT_REFRESH_EXPIRY = process.env.JWT_REFRESH_EXPIRY || '365d';

export interface JwtPayload {
  email: string;
  type?: string;
  iat?: number;
  exp?: number;
}

export function generateAccessToken(email: string): string {
  return jwt.sign({ email }, JWT_SECRET, { expiresIn: JWT_EXPIRY } as jwt.SignOptions);
}

export function generateRefreshToken(email: string): string {
  return jwt.sign({ email, type: 'refresh' }, JWT_SECRET, { expiresIn: JWT_REFRESH_EXPIRY } as jwt.SignOptions);
}

export function verifyToken(token: string): JwtPayload {
  return jwt.verify(token, JWT_SECRET) as JwtPayload;
}

export function decodeToken(token: string): JwtPayload | null {
  try {
    return jwt.decode(token) as JwtPayload;
  } catch {
    return null;
  }
}
